### ISSUE TITLE
Evaluating the boundaries after reading a checkpoint fails for some fields

### ISSUE THREAD

**Issue by MakisH** [precice/openfoam-adapter#4]
In OpenFOAM, each cell has a value at the center, but it can also have values at its faces, for example for boundary cells.

It has been found that re-evaluating the faces of the interface cells after reading a checkpoint can help the stability of the solution. This is now done for (almost) all the checkpointed fields of type `volScalarField` and `volVectorField`, by default.

However, it looks like the way this is implemented causes problems for some fields. In particular for turbulent solvers that register the `epsilon` field, the field is checkpointed, but when trying to evaluate the boundaries, it cannot find it. More strangely, it requests another field:

```
--> FOAM FATAL ERROR: 

    request for volScalarField::Internal kEpsilon:G from objectRegistry region0 failed
    available objects of type volScalarField::Internal are

21
(
thermo:mu
thermo:psi
nut
K
K_0
h
p_rgh_0
rho
k
h_0
p_rgh
dpdt
gh
alphat
rho_0
p
T
p_0
thermo:rho
epsilon
thermo:alpha
)


    From function const Type& Foam::objectRegistry::lookupObject(const Foam::word&) const [with Type = Foam::DimensionedField<double, Foam::volMesh>]
    in file /home/ubuntu/OpenFOAM/OpenFOAM-5.x/src/OpenFOAM/lnInclude/objectRegistryTemplates.C at line 193.
```

The methods involved are the `preciceAdapter::Adapter::readCheckpoint()` and the `preciceAdapter::Adapter::setupCheckpointing()`, which are implemented in the [Adapter.C](https://github.com/precice/openfoam-adapter/blob/master/Adapter.C).

The object `epsilon` is a match when executing `mesh_.lookupClass<volScalarField>().toc()`.

Currently, no object named `epsilon` is re-evaluated, in order to prevent unexpected errors. A warning is shown when adding fields to the list of checkpointed fields.

**In order to reproduce this bug**, do the following:
1. In the [Adapter.C](https://github.com/precice/openfoam-adapter/blob/master/Adapter.C), in `preciceAdapter::Adapter::readCheckpoint()`, comment the `if ("epsilon" != volScalarFields_.at(i)->name())` and the two brackets.
2. In the tutorial, edit the file [Fluid/constant/turbulenceProperties](https://github.com/precice/openfoam-adapter/blob/master/tutorials/CHT/flow-over-plate/buoyantPimpleFoam-laplacianFoam/Fluid/constant/turbulenceProperties) and set `simulationType` to `RAS` (`RASModel` should be set to `kEpsilon` already).
3. Run the tutorial starting the scripts `runFluid` and `runSolid` in different terminals. The fluid participant should crash when reading a checkpoint.
Make sure to have built the adapter with `-DADAPTER_DEBUG_MODE`, in order to see the debug messages.

**Comment by MakisH** [comment]
Hi! As a workaround, please set `evaluateBoundaries: No` in your adapter's configuration file. [Read more here](https://github.com/precice/openfoam-adapter/wiki/Configuration#debugging).

**Comment by deliobs** [comment]
I can't find the file Adapter.C. Is there a common path for this file? Is there a specific path for each case?

**Comment by MakisH** [comment]
The file I mean is the `precice-adapter-config.yml` of your case. If you still are looking for `Adapter.C`, it is located in the top-level directory of the source.

**Comment by deliobs** [comment]
I can' find both. Please, indicate entire direction. When you say: "directory of the source", are you referring to the path $FOAM_SRC?. Probably my trouble is a little bit different, even I will provide message out from terminal when tried simlulate.
`Create time

Create mesh for time = 0


PISO: Operating solver in PISO mode

Reading field p

Reading field U

Reading/calculating face flux field phi

AMI: Creating addressing and weights between 12968 source faces and 3950 target faces
AMI: Patch source sum(weights) min/max/average = 0.985939, 1, 0.999748
AMI: Patch target sum(weights) min/max/average = 1, 1, 1
AMI: Creating addressing and weights between 83738 source faces and 20868 target faces
AMI: Patch source sum(weights) min/max/average = 0.999864, 1.00045, 1
AMI: Patch target sum(weights) min/max/average = 1.00002, 1.00015, 1.00005
AMI: Creating addressing and weights between 13008 source faces and 3251 target faces
AMI: Patch source sum(weights) min/max/average = 0.985815, 1, 0.999749
AMI: Patch target sum(weights) min/max/average = 0.894072, 1, 0.999722
Selecting incompressible transport model Newtonian
Selecting turbulence model type RAS
Selecting RAS turbulence model kEpsilon
kEpsilonCoeffs
{
    Cmu             0.09;
    C1              1.44;
    C2              1.92;
    C3              -0.33;
    sigmak          1;
    sigmaEps        1.3;
}

Creating MRF zone list from MRFProperties
    creating MRF zone: MRF1
No finite volume options present


Starting time loop

Time = 1e-05

Courant Number mean: 0 max: 0
DILUPBiCCCG:  Solving for Ux, Initial residual = 1, Final residual = 3.78458e-09, No Iterations 1
DILUPBiCCCG:  Solving for Uy, Initial residual = 1, Final residual = 3.90832e-09, No Iterations 1
DILUPBiCCCG:  Solving for Uz, Initial residual = 1, Final residual = 1.14825e-10, No Iterations 1


--> FOAM FATAL ERROR: 

    request for volScalarField::Internal kEpsilon:G from objectRegistry region0 failed
    available objects of type volScalarField::Internal are

7
(
nut
k
(1|A(U))
nu
p
div(phiHbyA)
epsilon
)


    From function const Type& Foam::objectRegistry::lookupObject(const Foam::word&) const [with Type = Foam::DimensionedField<double, Foam::volMesh>]
    in file /home/ubuntu/OpenFOAM/OpenFOAM-4.1/src/OpenFOAM/lnInclude/objectRegistryTemplates.C at line 193.

FOAM aborting

#0  Foam::error::printStack(Foam::Ostream&) at ??:?
#1  Foam::error::abort() at ??:?
#2  Foam::DimensionedField<double, Foam::volMesh> const& Foam::objectRegistry::lookupObject<Foam::DimensionedField<double, Foam::volMesh> >(Foam::word const&) const at ??:?
#3  Foam::epsilonWallFunctionFvPatchScalarField::updateCoeffs() at ??:?
#4  Foam::GeometricField<double, Foam::fvPatchField, Foam::volMesh>::Boundary::updateCoeffs() at ??:?
#5  Foam::fvMatrix<double>::fvMatrix(Foam::GeometricField<double, Foam::fvPatchField, Foam::volMesh> const&, Foam::dimensionSet const&) at ??:?
#6  Foam::fv::gaussLaplacianScheme<double, Foam::SymmTensor<double> >::fvmLaplacianUncorrected(Foam::GeometricField<double, Foam::fvsPatchField, Foam::surfaceMesh> const&, Foam::GeometricField<double, Foam::fvsPatchField, Foam::surfaceMesh> const&, Foam::GeometricField<double, Foam::fvPatchField, Foam::volMesh> const&) at ??:?
#7  Foam::fv::gaussLaplacianScheme<double, double>::fvmLaplacian(Foam::GeometricField<double, Foam::fvsPatchField, Foam::surfaceMesh> const&, Foam::GeometricField<double, Foam::fvPatchField, Foam::volMesh> const&) at ??:?
#8  Foam::fv::laplacianScheme<double, double>::fvmLaplacian(Foam::GeometricField<double, Foam::fvPatchField, Foam::volMesh> const&, Foam::GeometricField<double, Foam::fvPatchField, Foam::volMesh> const&) at ??:?
#9  Foam::tmp<Foam::fvMatrix<double> > Foam::fvm::laplacian<double, double>(Foam::GeometricField<double, Foam::fvPatchField, Foam::volMesh> const&, Foam::GeometricField<double, Foam::fvPatchField, Foam::volMesh> const&, Foam::word const&) at ??:?
#10  ? at ??:?
#11  __libc_start_main in "/lib/x86_64-linux-gnu/libc.so.6"
#12  ? at ??:?
`

**Comment by MakisH** [comment]
If you have your simulation case files in e.g. `$FOAM_RUN/my_simulation/Fluid` and `$FOAM_RUN/my_simulation/Solid`, then you need the file `$FOAM_RUN/my_simulation/Fluid/precice-adapter-config.yml` ([example](https://github.com/precice/openfoam-adapter/blob/master/tutorials/CHT/flow-over-plate/buoyantPimpleFoam-laplacianFoam/Fluid/precice-adapter-config.yml)). Add `evaluateBoundaries: No` at the end of it.

**Comment by deliobs** [comment]
I just have in my case one path in my simulation. Because I am performing na MRF I must merge two merges in on folder, so I have two regions (0 and 1), but I don’t have two paths. Is this going to work in this situation?

**Comment by MakisH** [comment]
I am not sure about your setup, but it should probably work, maybe with some adjustments in the configuration. We can see this later.

But for the moment you are already starting the solvers and you run into this error, so we need to solve this first. You must have a `precice-adapter-config.yml` file somewhere and each solver needs its own one.

I am not sure what you mean by merging two cases. Do you run one MRF OpenFOAM solver (with heat transfer) and another solver in addition to this? Where do you have your `system/controlDict` file? The `precice-adapter-config.yml` should be in the parent directory of the `system/`.

**Comment by deliobs** [comment]
When I say about merging two is meaning merging boundaries regions in each mesh ( MRF region and Stator one). I prepared two meshes, after I used the command: mergeMeshes. So I got one mesh with all boundaries regions that I need.

Concerning quantities of solvers used. That’s only one (pisoFoam). I used it coupled to a turbulence model (kEpsilon) for solving a fluid flowing inside a mixing vessel. After setting properly boundaries conditions I got that mentioned message out from terminal.


________________________________

**Comment by MakisH** [comment]
I am not sure I understand the structure of your case. However, in order to solve this problem, you need to add `evaluateBoundaries: No` into your `precice-adapter-config.yml`. When you created your case, you must have created this file, otherwise the adapter would complain. So, find and edit this file.

Let's keep the discussion here only about the specific issue. For any other problems, please use our mailing list or Gitter. :)

**Comment by MakisH** [comment]
Very useful hint on this from @derekrisseeuw:

> Maybe this is because epsilon is usually coupled with wall functions. Other turbulent properties also generally make use of wallfunctions, so this list might need more entries.

**Comment by MakisH** [comment]
Additional entries we could consider, according to the experience of @moaxm: `epsilon_0`, `omega`, `omega_0`, `cellDisplacementx`, `cellDisplacementy`, `cellDisplacementz`.

**Comment by derekrisseeuw** [comment]
> Additional entries we could consider, according to the experience of @moaxm: `epsilon_0`, `omega`, `omega_0`, `cellDisplacementx`, `cellDisplacementy`, `cellDisplacementz`.

This is interesting. Maybe it's also a good idea to evaluate `k`, and `k_0`, since these are also turbulent properties and work with wall models. 

On a side node, the `k`, `omega` and `epsilon` require solving of another transport equation with a time derivative of the respective quantity. It could be that the `x_00` are called too during the evaluation of the boundary, but I am not sure of that. @moaxm what timescheme did you used and discovered these problems?

In OF-1712 there is the cellDisplacement vector field. I don't think this ever needs evaluation, since it is assigned by interpolation of the pointDisplacement field. (for displacement mapping).

**Comment by augustinhugues** [comment]
I have a case with a k-OmegaSST turbulence model (it looks like the k-epsilon one), I get the same error, OpenFoam is looking for a kOmegaSST field which does not exist. Using the same workaround `if ("omega" != volScalarFields_.at(i)->name())`  the calculation works with **parallel explicit coupling**, even if it is computing, I am not sure about the accuracy of the force. 
If I try **parallel implicit coupling** to improve stability, but I get an error because of different dimensions:
`
[1] Different dimensions for =
     dimensions : [0 1 -1 0 0 0 0] = [0 0 -1 0 0 0 0]
`

**Comment by MakisH** [comment]
@augustinhugues thank you for the information, I will try to find a better way to implement this functionality and solve the issue. Until then, please use `evaluateBoundaries: No` in your config.

Apart from this, you only get the error in implicit coupling, because only implicit coupling needs checkpointing. Again, the same workaround should work.

**Comment by DavidSCN** [comment]
We use the function `correctBoundaryConditions()` to reevaluate the boundaries e.g. for `zeroGradient` or `slip` boundary conditions, the boundary values change over time. For `fixedValue` the values are time independent and don't change with time. In our case, the error originates from applied wall functions (epsilonWallFunction). If you change the boundary conditions to more common types e.g. `zeroGradient` this error does not occur.

Calling the function `correctBoundaryConditions()` (`evaluate`) for wall functions leads to this error message and it seems like this is not necessary at all. There is also a [comment in an older OpenFOAM version](https://github.com/OpenFOAM/OpenFOAM-2.4.x/blob/2b147f41daf9ca07d0fb4c6b0576dc3d10a435f3/applications/utilities/preProcessing/applyBoundaryLayer/applyBoundaryLayer.C#L174), which confirms this in my opinion. The function `correctBoundaryConditions()`  is also not used in any wall function. Assuming it should be skipped, the problem would be: We only want to correct patches with wall function boundary condition.

**Comment by DavidSCN** [comment]
First step here: Find out about the specific boundary condition type on each patch. This can be done in the `Adapter.C` like this:
```                
wordList BoundaryTypes = volScalarFields_.at(i)->boundaryField().types();
```
It returns a wordList containing all necessary information. 

In the next step we need to correct the boundary conditions on specific types:
There is unfortunately no overloaded version of `GeometricField::correctBoundaryConditions()`, which works on patches instead of boundaries. 
Another approach would be to apply directly code contained in this function in the adapter. This is mainly the `evaluate()` or `correct()` function. But even the trial to access specific patches on each field is invalid. I got stuck at this point. Maybe someone else has an idea, how to treat this problem.

**Comment by DavidSCN** [comment]
After some investigations I sum up the current findings:

As already stated, `correctBoundaryConditions()` affects only certain BCs e.g. `zeroGradient` or some wall functions. Not every field, which uses these BCs, is modified, when `correctBoundaryConditions()` is called e.g. in a laminar CHT case only the field `p_rgh` is actually modified. 

Now the overall question is: How is the field/data modified?
If we reload a checkpoint, we want to have exactly the same values as before we started our calculations for the current timestep. Hence, we have:
```
checkpointData
calculateTimeStep()
reloadData
correctBoundaryConditions()
finalReloadData
```
We want the `finalReloadData` to be the same as the `checkpointData`. After looking at all (checkpointed) fields for a laminar and turbulent CHT case, the conclusion is, that data is reloaded correctly in the adapter and in any case, where `correctBoundaryConditions()` manipulates data, it becomes 'wrong' i.e. it does not correspond to the checkpointData any more.

Therefore, I propose we delete the code line(s), since they are not needed.

**Comment by MakisH** [comment]
This is a great summary! Thank you, @DavidSCN! Could you also add here some example results with and without the `correctBoundaryConditions()`? Maybe also with a small code snippet with them.

**Comment by DavidSCN** [comment]
Let me try:

i) at first, we look at the values during the checkpoint writing:
```
void preciceAdapter::Adapter::writeCheckpoint()
{
    for (uint i = 0; i < volScalarFields_.size(); i++)
    {
        *(volScalarFieldCopies_.at(i)) == *(volScalarFields_.at(i));

        Info<<"Checkpoint values: "<<endl;
        forAll(volScalarFields_.at(i)->boundaryField(), patchI)
            forAll(volScalarFields_.at(i)->boundaryField()[patchI],faceI)
                std::cout<<volScalarFields_.at(i)->name()<<"  "<<patchI<<"  "<<faceI<<"  "<<volScalarFields_.at(i)->boundaryField()[patchI][faceI]<<std::setprecision(15)<<std::endl;
    }
...
```
ii) Then we look at the calculated values (not necessary). std::cout is always the same and will be skipped:
```
  // Reload all the fields of type volScalarField
    for (uint i = 0; i < volScalarFields_.size(); i++)
    { 
        Info<<"Calculated values: "<<endl;
        forAll(volScalarFields_.at(i)->boundaryField(), patchI)
            forAll(... 

        *(volScalarFields_.at(i)) == *(volScalarFieldCopies_.at(i));
```

iii) Then we look at the reloaded values and the corrected values:
```
 if (evaluateBoundaries_)

            try{
                Info<<"Reloaded values: "<<endl;
                forAll(volScalarFields_.at(i)->boundaryField(), patchI)
                        forAll(...

                volScalarFields_.at(i)->correctBoundaryConditions();

                Info<<"Corrected values: "<<endl;
                forAll(volScalarFields_.at(i)->boundaryField(), patchI)
                        forAll(...
```

The pictures shows some random samples of the first iteration (not that nice, but hopefully enough):
![nut_example](https://user-images.githubusercontent.com/33414590/74163572-dab33c80-4c22-11ea-9957-d7d8ed8158dd.png)
![k_example](https://user-images.githubusercontent.com/33414590/74163932-5f9e5600-4c23-11ea-864c-74c57cf1fb05.png)
![p_rgh_example](https://user-images.githubusercontent.com/33414590/74164392-008d1100-4c24-11ea-8cab-219cc6904a6b.png)

**Comment by MakisH** [comment]
Looks good! Could you also check for `alphat` please? This was my initial motivation:

https://github.com/ludcila/CHT-preCICE/blob/9943392fca4ab3548576f7e34899cd8f7159c232/solvers/OpenFOAM/solvers/buoyantPimpleFoam/buoyantPimpleFoam.C#L211

**Comment by DavidSCN** [comment]
Done with the default `compressible::alphatWallFunction`, in the first iteration, there are just `k` and `p_rgh` (and `epsilon` fails), who are affected by the function. 

Do we (Lucia) know for sure, that the code line improved/changed the results? Or just an assumption? Does she maybe uses other BCs?

**Comment by MakisH** [comment]
> Do we (Lucia) know for sure, that the code line improved/changed the results? Or just an assumption? Does she maybe uses other BCs?

From [Lucia's thesis, page 37](http://www5.in.tum.de/pub/Cheung2016_Thesis.pdf):
> The reason why all the checkpointing related code is not moved into the adapter (e.g. the en-
tire if(adapter.isReadCheckpointRequired()) clause), is that there may be solver-specific
actions that need to be included in the checkpointing. For example, in the case of buoyant-
PimpleFoam it was found that the method correctBoundaryConditions() of the turbulent
thermal diffusivity alphat had to be called, otherwise, the convergence of the coupling would
be affected.

**Comment by MakisH** [comment]
Have been closed with #110.
